modules/rtpengine/rtpengine.c

```diff
@@ -2191,81 +2272,176 @@ static struct rtpp_set * select_rtpp_set(int id_set ){
+	if (shm_str_dup(&entry->callid, &callid) < 0) {
+		LM_ERR("rtpengine hash table fail to duplicate calllen=%d callid=%.*s",
+			callid.len, callid.len, callid.s);
+	}
+	entry->node = node;
+	entry->next = NULL;
+	entry->tout = get_ticks() + hash_table_tout;
+
+	/* Insert the key<->entry from the hashtable */
+	if (!rtpengine_hash_table_insert(&callid, entry)) {
+		LM_ERR("rtpengine hash table fail to insert node=%.*s for calllen=%d callid=%.*s",
+			node->rn_url.len, node->rn_url.s, callid.len, callid.len, callid.s);
+	} else {
+		LM_DBG("rtpengine hash table insert node=%.*s for calllen=%d callid=%.*s\n",
+			node->rn_url.len, node->rn_url.s, callid.len, callid.len, callid.s);
+	}
+
+	/* Return selected node  */
 	return node;
 }
 
+/*
+ * lookup the hastable (key=callid value=node) and get the old node (e.g. for answer/delete)
+ */
+static struct rtpp_node *
+select_rtpp_node_old(str callid, int do_test, int op)
+{
+	struct rtpp_node *node = NULL;
+	struct rtpengine_hash_entry *entry = NULL;
+
+	entry = rtpengine_hash_table_lookup(&callid);
+	if (!entry) {
+		LM_ERR("rtpengine hash table lookup failed to find entry for calllen=%d callid=%.*s\n",
+			callid.len, callid.len, callid.s);
+	} else {
+		LM_DBG("rtpengine hash table lookup find entry for calllen=%d callid=%.*s\n",
+			callid.len, callid.len, callid.s);
+		node = entry->node;
+	}
+
+	if (!node) {
+		LM_ERR("rtpengine hash table lookup failed to find node for calllen=%d callid=%.*s\n",
+			callid.len, callid.len, callid.s);
+		return NULL;
+	} else {
+		LM_DBG("rtpengine hash table lookup find node=%.*s for calllen=%d callid=%.*s\n",
+			node->rn_url.len, node->rn_url.s, callid.len, callid.len, callid.s);
+	}
+
+	// if node enabled, return it
+	if (!node->rn_disabled) {
+		return node;
+	}
+
+	// if node _manually_ disabled(e.g kamctl) and proper configuration, return it
+	if (node->rn_recheck_ticks == MI_MAX_RECHECK_TICKS) {
+		if (rtpengine_allow_op) {
+			LM_DBG("node=%.*s for calllen=%d callid=%.*s is disabled(permanent) (probably still UP)! Return it\n",
+				node->rn_url.len, node->rn_url.s, callid.len, callid.len, callid.s);
+			return node;
+		}
+		LM_DBG("node=%.*s for calllen=%d callid=%.*s is disabled(permanent) (probably still UP)! Return NULL\n",
+			node->rn_url.len, node->rn_url.s, callid.len, callid.len, callid.s);
+	} else {
+		LM_DBG("node=%.*s for calllen=%d callid=%.*s is disabled (probably BROKE)! Return NULL\n",
+			node->rn_url.len, node->rn_url.s, callid.len, callid.len, callid.s);
+	}
+
+	return NULL;
+}
+
+/*
+ * Main balancing routine. This DO try to keep the same proxy for
+ * the call if some proxies were disabled or enabled (e.g. kamctl command)
+ */
+static struct rtpp_node *
+select_rtpp_node(str callid, int do_test, int op)
+{
+	if(!active_rtpp_set) {
+		LM_ERR("script error - no valid set selected\n");
+		return NULL;
+	}
+
+	// calculate and choose a node
+	if (op == OP_OFFER) {
+		// run the selection algorithm
```

Comment:
"Offer" doesn't necessarily mean that this will be a new entry. A re-invite will also do an "offer" but needs to do a lookup into the hash table first. I think a lookup should always be done first, with fallback to creating a new entry in case of an offer.